test/e2e/leaderelection.go

```diff
@@ -0,0 +1,104 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/leaderelection"
+
+	app "github.com/kubernetes-sigs/federation-v2/cmd/controller-manager/app/leaderelection"
+	"github.com/kubernetes-sigs/federation-v2/cmd/controller-manager/app/options"
+	"github.com/kubernetes-sigs/federation-v2/pkg/controller/util"
+	"github.com/kubernetes-sigs/federation-v2/test/common"
+	"github.com/kubernetes-sigs/federation-v2/test/e2e/framework"
+	. "github.com/onsi/ginkgo"
+)
+
+var _ = Describe("Leader Elector", func() {
+	f := framework.NewFederationFramework("leaderelection")
+	tl := framework.NewE2ELogger()
+
+	var leaderElection *util.LeaderElectionConfiguration
+	var opts *options.Options
+
+	BeforeEach(func() {
+		leaderElection = &util.LeaderElectionConfiguration{
+			LeaderElect:   true,
+			LeaseDuration: time.Second,
+			RenewDeadline: 500 * time.Millisecond,
+			RetryPeriod:   100 * time.Millisecond,
+			ResourceLock:  "configmaps",
+		}
+
+		opts = options.NewOptions()
+		opts.Config = f.ControllerConfig()
+		opts.LeaderElection = leaderElection
+	})
+
+	It("should chose secondary instance, primary goes down", func() {
+		if !framework.TestContext.TestManagedFederation {
+			framework.Skipf("leader election is valid only in managed federation setup")
+		}
+
+		primaryLeaderElector, err := app.LeaderElector(opts, func(opts *options.Options, stopChan <-chan struct{}) {
+			tl.Log("Run controllers of primary controller manager")
+			<-stopChan
+		})
+		framework.ExpectNoError(err)
+
+		ctx := context.Background()
+		primaryContext, primaryContextCancel := context.WithCancel(ctx)
+
+		tl.Log("Running primary instance of controller manager")
+		go primaryLeaderElector.Run(primaryContext)
+		waitToBecomeLeader(tl, primaryLeaderElector)
+		tl.Log("Primary instance is elected leader")
+
+		secondaryLeaderElector, err := app.LeaderElector(opts, func(opts *options.Options, stopChan <-chan struct{}) {
+			tl.Log("Run controllers of secondary controller manager")
+			<-stopChan
+		})
+		framework.ExpectNoError(err)
+
+		SecondaryContext, secondaryContextCancel := context.WithCancel(ctx)
+		tl.Log("Running secondary instance of controller manager")
+		go secondaryLeaderElector.Run(SecondaryContext)
+
+		// Stop primary instance of controller manager
+		primaryContextCancel()
+
+		waitToBecomeLeader(tl, secondaryLeaderElector)
+		tl.Log("Secondary instance is elected leader")
+
+		secondaryContextCancel()
```

Comment:
[Not necessary for this PR merge, as these scenario might be tested somewhere else, I did not find them in client-go though] I think additional useful tests might be:
1 - Both instances start together, only one of them becomes the leader
2 - Emulate network partitioning, both become leaders, when communication is back, one of them relents leadership

Reply:
both become leaders? isn't one of network partitions with etcd quorum expected to win?

Reply:
@yamt, My bad. I did assume this for the controllers running in two different clusters (without looking at the internals of the implementation). You are right that both controllers will need to be synced up to an API server (or multiple API servers) backed by same quorum backed store.

Reply:
Likely not a solved problem anytime soon for arbitrary environments.  There's a reason raft has a hard upper bound on latency between nodes - that's inherent complexity that will be hard or impossible to solve.